---
title: "Multisequence alighment"
author: "Rafael Villasmil"
date: "10/16/2021"
output: pdf_document
---

```{r install-packages, echo=FALSE}
install.packages("ape")
install.packages("seqinr")
```

```{r load-packages}
library(ape)
#library(seqinr)
library(Biostrings)
library(msa)
library(ggpubr)
```

```{r read-fasta-files-from-ncbi}
```

```{r load_seq_of_paper_species}
setwd("~/GitHub/cd4-aligment-nj/speciens_in_paper")
cd4_stringset_paper <- readDNAStringSet(list.files(pattern = "fa"))
cd4_ClustalW_paper <- msa(cd4_stringset_paper, method="ClustalW") 
working_clustalW <- cd4_ClustalW_paper
rownames(working_clustalW) <- c( "Mouse","Norway rat", "Human","Chimpanzee", "Rhesus monkey","Pig","Cattle", "Dog","Rabbit", "Chicken")
autoMasked <- maskGaps(working_clustalW, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")
cd4_nj <- nj(sdist)
plot(cd4_nj)
```


```{r explore cd4_ClustalW}
rownames(cd4_ClustalW)
rownames(working_clustalW) <- c("Human", "Chimpanzee", "Rhesus monkey", 
                                "Green Monkey", "Cat", "Dog", "Mouse",
                                "Norway rat", "Pig", "Cattle")
detail(cd4_ClustalW)
maskTest <- cd4_ClustalW
rowmask(maskTest) <- IRanges(start=1,end=3)
rowmask(maskTest)
maskTest
```

use Multiple Aligment Objects 
```{r distance_matrix_calculation}
autoMasked <- maskGaps(working_clustalW, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")


clust <- hclust(sdist, method = "single")
plot(clust)

cd4_nj <- nj(sdist)
plot(cd4_nj)
```


https://stackoverflow.com/questions/35016713/how-do-i-import-a-jsonl-file-in-r-and-how-do-i-transform-it-in-csv
```{r}
library(jsonlite)
library(dplyr)

list_directories <- 
  list.dirs(path = "C:/Users/villasmilr/Documents/GitHub/cd4-aligment-nj/data")
jsnol_directories <- list_directories[seq(3, 41, by = 2)]

table_1all <- matrix(ncol = 8)
colnames(table_1all) <- c("Species", "Chromosome", "Gene_length", "No.of_exon", 
                      "Exon(bp)","Coding_region_(bp)", "Intron(bp)", "UTR(bp)")
table_1all <- as.data.frame(table_1all)

 for (i in 1:length(jsnol_directories)){
  lines <- readLines(paste(jsnol_directories[i], "/data_report.jsonl", sep= ""))
  lines <- lapply(lines, fromJSON)
  lines <- lapply(lines, unlist)
  data_report <- bind_rows(lines)
  
  gene <- readDNAStringSet(paste(jsnol_directories[i], "/gene.fna", sep= ""))
  protein <- 
    readAAStringSet(paste(jsnol_directories[i], "/protein.faa", sep= ""))
  rna <- readRNAStringSet(paste(jsnol_directories[i], "/rna.fna", sep= ""))
  data_table <- read.delim(paste(jsnol_directories[i], "/data_table.tsv", 
                                 sep= ""), 
                           header = TRUE, stringsAsFactors = FALSE, quote = "", 
                           sep = "\t")
  
  table_1all[i,"Species"] <- data_report$taxname
  table_1all[i,"Chromosome"] <- data_report$chromosomes
  table_1all[i,"Exon(bp)"] <- data_table$transcript_length[1]
  table_1all[i,"Gene_length"] <- gene@ranges@width[1]
  table_1all[i,"Coding_region_(bp)"] <- rna@ranges@width[1]
  
}

list_species <- c("Bos taurus",              "Canis lupus familiaris", 
                  "Chlorocebus sabaeus",     "Gallus gallus",         
                  "Gorilla gorilla",         "Homo sapiens",           
                  "Hylobates moloch",        "Macaca mulatta",         
                  "Macaca nemestrina",       "Monodelphis domestica",  
                  "Mus musculus",            "Nomascus leucogenys",    
                  "Oryctolagus cuniculus",   "Pan paniscus",           
                  "Pan troglodytes",         "Pongo abelii",           
                  "Rattus norvegicus",       "Rhinopithecus bieti",    
                  "Rhinopithecus roxellana", "Sus scrofa")

No.of_exon <- c(11, 11,
                10, 11,
                11, 11,
                10, 11,
                12, 14,
                10, 10,
                10, 10,
                11, 10,
                11, 11,
                11, 10)

table_1all$No.of_exon <- No.of_exon
table_1all$`Intron(bp)` <- table_1all$Gene_length - table_1all$`Exon(bp)`
table_1all$`UTR(bp)` <- table_1all$`Exon(bp)` - table_1all$`Coding_region_(bp)`

list_ori_species <- c("Homo sapiens", "Pan troglodytes",
                      "Macaca mulatta", "Canis lupus familiaris", 
                      "Bos taurus","Sus scrofa",
                      "Monodelphis domestica", "Oryctolagus cuniculus",
                      "Mus musculus","Rattus norvegicus", "Gallus gallus")

list_primate_species <- c("Homo sapiens", "Chlorocebus sabaeus",            
                          "Gorilla gorilla", "Hylobates moloch",                 
                          "Macaca mulatta", "Macaca nemestrina",       
                          "Nomascus leucogenys", "Pan paniscus",           
                          "Pan troglodytes", "Pongo abelii", 
                          "Rhinopithecus bieti","Rhinopithecus roxellana")


# make table 1 with original species 
table_1ori <- matrix(NA, ncol = ncol(table_1all), nrow =0 )
colnames(table_1ori) <- colnames(table_1all)
for (i in 1:length(list_ori_species)){
  table_1ori <- 
    rbind(table_1ori, 
          table_1all[which(grepl(list_ori_species[i], table_1all$Species)),])
}
write.csv(table_1ori,"~/GitHub/cd4-aligment-nj/Table 1 orginal species.csv", 
          row.names = TRUE)

# make table 1 with new species 
table_1primates <- matrix(NA, ncol = ncol(table_1all), nrow =0 )
colnames(table_1primates) <- colnames(table_1all)
for (i in 1:length(list_ori_species)){
  table_1primates <- 
    rbind(table_1primates, 
          table_1all[which(grepl(list_primate_species[i], 
                                 table_1all$Species)),])
}
write.csv(table_1primates,
          "~/GitHub/cd4-aligment-nj/Table 1 primates.csv", row.names = TRUE)
```


```{r paper_comparisons}
# minimun gene length 
table_1all[which(grepl(min(table_1ori$Gene_length), 
                                 table_1all$Gene_length)),]

# maximun gene length 
table_1all[which(grepl(max(table_1ori$Gene_length), 
                                 table_1all$Gene_length)),]
```



```{r correlation between }
# original assertion 
ggscatter(table_1ori, x = "Gene_length", y = "Intron(bp)", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Gene Length", ylab = "Intron (bp)") + 
  labs(title = "Original Species With Data as of Nov 2021")

# exon in original species
ggscatter(table_1ori, x = "Gene_length", y = "Exon(bp)", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Gene Length", ylab = "Exon (bp)") + 
  labs(title = "Original Species With Data as of Nov 2021")

# what about new data 
ggscatter(table_1primates, x = "Gene_length", y = "Intron(bp)", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Gene Length", ylab = "Intron (bp)") + 
  labs(title = "Primates With Data as of Nov 2021")

ggscatter(table_1primates, x = "Gene_length", y = "Exon(bp)", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Gene Length", ylab = "Exon (bp)") + 
  labs(title = "Primates With Data as of Nov 2021")
```

```{r aling_and_build_phylo_tree_with_data_in_data}
setwd("~/GitHub/cd4-aligment-nj/data/dna_original")
gene_ori <- readDNAStringSet(list.files(pattern = "fna"))
# clustalW_ori <- msa(gene_ori, method="ClustalW") 
autoMasked <- maskGaps(clustalW_ori, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")
nj_ori <- nj(sdist)
nj_ori$tip.label <- sub(".*organism=","",nj_ori$tip.label)
plot(nj_ori)

setwd("~/GitHub/cd4-aligment-nj/data/dna_primates")
gene_primates <- readDNAStringSet(list.files(pattern = "fna"))
# clustalW_ori <- msa(gene_primates, method="ClustalW") 
autoMasked <- maskGaps(clustalW_ori, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")
nj_primates <- nj(sdist)
plot(nj_primates)
```



```{r protr_package}
library(protr)

setwd("~/GitHub/cd4-aligment-nj/data")
cd4_seq <- c(list.files(pattern = "fa"))
cd4_dna <- read.dna(list.files(pattern = "fa"))
```

