---
title: "Multisequence alighment"
author: "Rafael Villasmil"
date: "10/16/2021"
output: pdf_document
---

```{r install-packages, echo=FALSE}
install.packages("ape")
install.packages("seqinr")
```

```{r load-packages}
library(ape)
#library(seqinr)
library(Biostrings)
library(msa)
```

```{r read-fasta-files-from-ncbi}
```

```{r load_seq_of_paper_species}
setwd("~/GitHub/cd4-aligment-nj/speciens _in_paper")
cd4_stringset_paper <- readDNAStringSet(list.files(pattern = "fa"))
cd4_stringset_paper <- readDNAStringSet(list.files(pattern = "fa"))
cd4_stringset_paper@ranges@NAMES
cd4_stringset_paper@ranges@width
cd4_ClustalW_paper <- msa(cd4_stringset_paper, method="ClustalW") 
working_clustalW <- cd4_ClustalW_paper
rownames(working_clustalW) <- c( "Mouse","Norway rat", "Human","Chimpanzee", "Rhesus monkey","Pig","Cattle", "Dog","Rabbit", "Chicken")
autoMasked <- maskGaps(working_clustalW, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")
cd4_nj <- nj(sdist)
plot(cd4_nj)
```


```{r explore cd4_ClustalW}
rownames(cd4_ClustalW)
rownames(working_clustalW) <- c("Human", "Chimpanzee", "Rhesus monkey", 
                                "Green Monkey", "Cat", "Dog", "Mouse",
                                "Norway rat", "Pig", "Cattle")
detail(cd4_ClustalW)
maskTest <- cd4_ClustalW
rowmask(maskTest) <- IRanges(start=1,end=3)
rowmask(maskTest)
maskTest
```

use Multiple Aligment Objects 
```{r distance_matrix_calculation}
autoMasked <- maskGaps(working_clustalW, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")


clust <- hclust(sdist, method = "single")
plot(clust)

cd4_nj <- nj(sdist)
plot(cd4_nj)
```


https://stackoverflow.com/questions/35016713/how-do-i-import-a-jsonl-file-in-r-and-how-do-i-transform-it-in-csv
```{r}
library(jsonlite)
library(dplyr)

list_directories <- 
  list.dirs(path = "C:/Users/villasmilr/Documents/GitHub/cd4-aligment-nj/data")
jsnol_directories <- list_directories[seq(3, 41, by = 2)]

table_1all <- matrix(ncol = 8)
colnames(table_1all) <- c("Species", "Chromosome", "Gene_length", "No.of_exon", 
                      "Exon(bp)","Coding_region_(bp)", "Intron(bp)", "UTR(bp)")
table_1all <- as.data.frame(table_1all)
 for (i in 1:length(jsnol_directories)){
  lines <- readLines(paste(jsnol_directories[i], "/data_report.jsonl", sep= ""))
  lines <- lapply(lines, fromJSON)
  lines <- lapply(lines, unlist)
  data_report <- bind_rows(lines)
  
  gene <- readDNAStringSet(paste(jsnol_directories[i], "/gene.fna", sep= ""))
  protein <- 
    readAAStringSet(paste(jsnol_directories[i], "/protein.faa", sep= ""))
  rna <- readRNAStringSet(paste(jsnol_directories[i], "/rna.fna", sep= ""))
  data_table <- read.delim(paste(jsnol_directories[i], "/data_table.tsv", 
                                 sep= ""), 
                           header = TRUE, stringsAsFactors = FALSE, quote = "", 
                           sep = "\t")
  
  table_1all[i,"Species"] <- data_report$taxname
  table_1all[i,"Chromosome"] <- data_report$chromosomes
  table_1all[i,"Exon(bp)"] <- data_table$transcript_length[1]
  table_1all[i,"Gene_length"] <- gene@ranges@width[1]
  table_1all[i,"Coding_region_(bp)"] <- rna@ranges@width[1]
  
}

list_species <- c("Bos taurus",              "Canis lupus familiaris", 
                  "Chlorocebus sabaeus",     "Gallus gallus",         
                  "Gorilla gorilla",         "Homo sapiens",           
                  "Hylobates moloch",        "Macaca mulatta",         
                  "Macaca nemestrina",       "Monodelphis domestica",  
                  "Mus musculus",            "Nomascus leucogenys",    
                  "Oryctolagus cuniculus",   "Pan paniscus",           
                  "Pan troglodytes",         "Pongo abelii",           
                  "Rattus norvegicus",       "Rhinopithecus bieti",    
                  "Rhinopithecus roxellana", "Sus scrofa")

No.of_exon <- c(11, 11,
                10, 11,
                11, 11,
                10, 11,
                12, 14,
                10, 10,
                10, 10,
                11, 10,
                11, 11,
                11, 10)

table_1all$
table_1all$`Intron(bp)` <- table_1all$Gene_length - table_1all$`Exon(bp)`
table_1all$`UTR(bp)` <- table_1all$`Exon(bp)` - table_1all$`Coding_region_(bp)`



```






```{r protr_package}
library(protr)

setwd("~/GitHub/cd4-aligment-nj/data")
cd4_seq <- c(list.files(pattern = "fa"))
cd4_dna <- read.dna(list.files(pattern = "fa"))
```

