---
title: "Multisequence alighment"
author: "Rafael Villasmil"
date: "10/16/2021"
output: pdf_document
---

```{r install-packages, echo=FALSE}
install.packages("ape")
install.packages("seqinr")
```

```{r load-packages}
library(ape)
library(Biostrings)
library(DECIPHER)
library(dplyr)
library(ggExtra)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(hrbrthemes)
library(msa)
library(reshape2)
library(RSQLite)
library(tidyverse)
library(jsonlite)
library(viridis)
```

```{r read-fasta-files-from-ncbi}
```

https://stackoverflow.com/questions/35016713/how-do-i-import-a-jsonl-file-in-r-and-how-do-i-transform-it-in-csv
```{r}
list_directories <- 
  list.dirs(path = "C:/Users/villasmilr/Documents/GitHub/cd4-aligment-nj/data")
jsnol_directories <- list_directories[seq(3, 41, by = 2)]

table_1all <- matrix(ncol = 8)
colnames(table_1all) <- c("Species", "Chromosome", "Gene_length", "No.of_exon", 
                      "Exon(bp)","Coding_region_(bp)", "Intron(bp)", "UTR(bp)")
table_1all <- as.data.frame(table_1all)

 for (i in 1:length(jsnol_directories)){
  lines <- readLines(paste(jsnol_directories[i], "/data_report.jsonl", sep= ""))
  lines <- lapply(lines, fromJSON)
  lines <- lapply(lines, unlist)
  data_report <- bind_rows(lines)
  
  gene <- readDNAStringSet(paste(jsnol_directories[i], "/gene.fna", sep= ""))
  protein <- 
    readAAStringSet(paste(jsnol_directories[i], "/protein.faa", sep= ""))
  rna <- readRNAStringSet(paste(jsnol_directories[i], "/rna.fna", sep= ""))
  data_table <- read.delim(paste(jsnol_directories[i], "/data_table.tsv", 
                                 sep= ""), 
                           header = TRUE, stringsAsFactors = FALSE, quote = "", 
                           sep = "\t")
  
  table_1all[i,"Species"] <- data_report$taxname
  table_1all[i,"Chromosome"] <- data_report$chromosomes
  table_1all[i,"Exon(bp)"] <- data_table$transcript_length[1]
  table_1all[i,"Gene_length"] <- gene@ranges@width[1]
  table_1all[i,"Coding_region_(bp)"] <- rna@ranges@width[1]
  
}

list_species <- test<- table_1all$Species

No.of_exon <- c(11, 11,
                10, 11,
                11, 11,
                10, 11,
                12, 14,
                10, 10,
                10, 10,
                11, 10,
                11, 11,
                11, 10)

table_1all$No.of_exon <- No.of_exon
table_1all$`Intron(bp)` <- table_1all$Gene_length - table_1all$`Exon(bp)`
table_1all$`UTR(bp)` <- table_1all$`Exon(bp)` - table_1all$`Coding_region_(bp)`

list_ori_species <- c("Homo sapiens", "Pan troglodytes",
                      "Macaca mulatta", "Canis lupus familiaris", 
                      "Bos taurus","Sus scrofa",
                      "Monodelphis domestica", "Oryctolagus cuniculus",
                      "Mus musculus","Rattus norvegicus", "Gallus gallus")

list_primate_species <- c("Homo sapiens", "Chlorocebus sabaeus",            
                          "Gorilla gorilla", "Hylobates moloch",              
                          "Macaca mulatta", "Macaca nemestrina",       
                          "Nomascus leucogenys", "Pan paniscus",           
                          "Pan troglodytes", "Pongo abelii", 
                          "Rhinopithecus bieti","Rhinopithecus roxellana")


# make table 1 with original species 
table_1ori <- matrix(NA, ncol = ncol(table_1all), nrow =0 )
colnames(table_1ori) <- colnames(table_1all)
for (i in 1:length(list_ori_species)){
  table_1ori <- 
    rbind(table_1ori, 
          table_1all[which(grepl(list_ori_species[i], table_1all$Species)),])
}
row.names(table_1ori) <- NULL
write.csv(table_1ori,"~/GitHub/cd4-aligment-nj/Table 1 orginal species.csv", 
          row.names = TRUE)

# make table 1 with new species 
table_1primates <- matrix(NA, ncol = ncol(table_1all), nrow =0 )
colnames(table_1primates) <- colnames(table_1all)
for (i in 1:length(list_primate_species)){
  table_1primates <- 
    rbind(table_1primates, 
          table_1all[which(grepl(list_primate_species[i], 
                                 table_1all$Species)),])
}
row.names(table_1primates) <- NULL
write.csv(table_1primates,
          "~/GitHub/cd4-aligment-nj/Table 1 primates.csv", row.names = TRUE)
```
```{r Figure_2_rendition}
#basic 
ggplot(data = table_1ori, 
       aes(x=`Coding_region_(bp)`, y=`Exon(bp)` , 
           size=Gene_length)) +
  geom_point() + 
  theme(legend.position = )

# original species
figure_2ori <- ggplot(data = table_1ori, 
       aes(x=`Coding_region_(bp)`, y=`Exon(bp)`, label = Species)) +
  geom_point(aes(size=Gene_length)) +
  xlim(1000, 3000) +
  labs(x = "Coding Region (bp)", y = "Exon (bp)", label = "Gene Length") +
  geom_text_repel(force = 10) + 
  theme(legend.position = c(0.9, 0.2))

figure_2ori
ggMarginal(figure_2ori, type="boxplot")

# primate species
figure_2primates <- ggplot(data = table_1primates, 
       aes(x=`Coding_region_(bp)`, y=`Exon(bp)`, label = Species)) +
  geom_point(aes(size=Gene_length)) +
  xlim(2200, 2700) + ylim(2800, 3400) +
  labs(x = "Coding Region (bp)", y = "Exon (bp)", label = "Gene Length") +
  geom_text_repel(force = 50) + 
  theme(legend.position = c(0.85, 0.3))

figure_2primates
ggMarginal(figure_2primates, type="boxplot")
```



```{r paper_comparisons}
# minimun gene length 
table_1all[which(grepl(min(table_1ori$Gene_length), 
                                 table_1all$Gene_length)),]

# maximun gene length 
table_1all[which(grepl(max(table_1ori$Gene_length), 
                                 table_1all$Gene_length)),]
```



```{r correlation between }
# original assertion 
ggscatter(table_1ori, x = "Gene_length", y = "Intron(bp)", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Gene Length", ylab = "Intron (bp)") + 
  labs(title = "Original Species With Data as of Nov 2021")

# exon in original species
ggscatter(table_1ori, x = "Gene_length", y = "Exon(bp)", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Gene Length", ylab = "Exon (bp)") + 
  labs(title = "Original Species With Data as of Nov 2021")

# what about new data 
ggscatter(table_1primates, x = "Gene_length", y = "Intron(bp)", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Gene Length", ylab = "Intron (bp)") + 
  labs(title = "Primates With Data as of Nov 2021")

ggscatter(table_1primates, x = "Gene_length", y = "Exon(bp)", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Gene Length", ylab = "Exon (bp)") + 
  labs(title = "Primates With Data as of Nov 2021")
```

```{r aling_and_build_phylo_tree_with_data_in_data, include = FALSE}
knitr::opts_chunk$set(eval = FALSE)

# build phylogenetic tree for species in original list
setwd("~/GitHub/cd4-aligment-nj/data_dna/dna_original")
gene_ori <- readDNAStringSet(list.files(pattern = "fna"))
clustalW_ori <- msa(gene_ori, method="ClustalW") 
autoMasked <- maskGaps(clustalW_ori, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")
nj_ori <- nj(sdist)
nj_ori$tip.label <- sub(".*organism=","",nj_ori$tip.label)
nj_ori$tip.label <- sub("(] ).*","",nj_ori$tip.label)
plot(nj_ori)

# build phylogenetic tree for primate species
setwd("~/GitHub/cd4-aligment-nj/data_dna/dna_primates")
gene_primates <- readDNAStringSet(list.files(pattern = "fna"))
clustalW_primates <- msa(gene_primates, method="ClustalW") 
autoMasked <- maskGaps(clustalW_primates, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")
nj_primates <- nj(sdist)
nj_primates$tip.label <- sub(".*organism=","",nj_primates$tip.label)
nj_primates$tip.label <- sub("(] ).*","",nj_primates$tip.label)
plot(nj_primates)
```

```{r secondary_structure_prediction}
setwd("~/GitHub/cd4-aligment-nj/data_protein/aa_orginial")
protein_ori <- readAAStringSet(list.files(pattern = "faa"))
predict_ori <- PredictHEC(protein_ori,
                          type = "states",
                          windowSize = 7,
                          background = c(H = -0.12, E = -0.25, C = 0.23),
                          HEC_MI1 = NULL,
                          HEC_MI2 = NULL)

# create table 2 with orginial species 
table_2ori <- data.frame(matrix(NA, ncol=4, 
                                nrow = length(protein_ori@ranges@NAMES)))
colnames(table_2ori) <- c("Species", "H", "E", "C")
for (i in 1:length(protein_ori@ranges@NAMES)){
table_2ori[i,"Species"] <- protein_ori@ranges@NAMES[i]
table_2ori[i,"C"] <- table(strsplit(predict_ori[i], ""))["C"]
table_2ori[i,"E"] <- table(strsplit(predict_ori[i], ""))["E"]
table_2ori[i,"H"] <- table(strsplit(predict_ori[i], ""))["H"]
}
# reduce Species to name
table_2ori$Species <- sub(".*organism=","",table_2ori$Specie)
table_2ori$Species <- sub("(] ).*","",table_2ori$Specie)
# reordet rows to match paper
reorder_idx <- match(list_ori_species, table_2ori$Species)
table_2ori <- table_2ori[reorder_idx,]
row.names(table_2ori) <- NULL
write.csv(table_2ori,
          "~/GitHub/cd4-aligment-nj/Table 2 original.csv", row.names = TRUE)


setwd("~/GitHub/cd4-aligment-nj/data_protein/aa_primates")
protein_primates <- readAAStringSet(list.files(pattern = "faa"))
predict_primates <- PredictHEC(protein_primates,
                               type = "states",
                               windowSize = 7,
                               background = c(H = -0.12, E = -0.25, C = 0.23),
                               HEC_MI1 = NULL,
                               HEC_MI2 = NULL)

# create table 2 with primate species 
table_2primates <- data.frame(matrix(NA, ncol=4, 
                                nrow = length(protein_primates@ranges@NAMES)))
colnames(table_2primates) <- c("Species", "H", "E", "C")
for (i in 1:length(protein_primates@ranges@NAMES)){
table_2primates[i,"Species"] <- protein_primates@ranges@NAMES[i]
table_2primates[i,"C"] <- table(strsplit(predict_primates[i], ""))["C"]
table_2primates[i,"E"] <- table(strsplit(predict_primates[i], ""))["E"]
table_2primates[i,"H"] <- table(strsplit(predict_primates[i], ""))["H"]
}
# reduce Species to name
table_2primates$Species <- sub(".*organism=","",table_2primates$Specie)
table_2primates$Species <- sub("(] ).*","",table_2primates$Specie)
# reordet rows to match paper
reorder_idx <- match(list_primate_species, table_2primates$Species)
table_2primates <- table_2primates[reorder_idx,]
row.names(table_2primates) <- NULL
write.csv(table_2primates,
          "~/GitHub/cd4-aligment-nj/Table 2 primates.csv", row.names = TRUE)

table_2ori_perc <- table_2ori
for (i in 1: nrow(table_2ori_perc)){
  for (j in 2:4){
  table_2ori_perc[i,j] <- table_2ori[i,j]/sum(table_2ori[i,2:4])
  }
}

table_2primates_perc <- table_2primates
for (i in 1: nrow(table_2primates_perc)){
  for (j in 2:4){
  table_2primates_perc[i,j] <- table_2primates[i,j]/sum(table_2primates[i,2:4])
  }
}
```

```{r plot_secondary_structure}
#original 
table_2subs_ori <-
  ggplot(melt(table_2ori_perc), aes(x=Species, y=value, fill=variable)) +
  geom_bar(position="fill", stat="identity", colour="black") +
  theme_bw() + 
  scale_fill_manual(values = c("grey10", "grey34", "grey67"),
                    breaks=c("H", "E", "C"),
                    labels=c("alpha Helix", "beta Sheet", "Coil")) +
  labs(x = NULL, y = "Fraction", fill = NULL) +
  coord_flip()

table_2subs_ori

table_2subs_primates <-
  ggplot(melt(table_2primates_perc), aes(x=Species, y=value, fill=variable)) +
  geom_bar(position="fill", stat="identity", colour="black") +
  theme_bw() + 
  scale_fill_manual(values = c("grey10", "grey34", "grey67"),
                    breaks=c("H", "E", "C"),
                    labels=c("alpha Helix", "beta Sheet", "Coil")) +
  labs(x = NULL, y = "Fraction", fill = NULL) +
  coord_flip()

table_2subs_primates
```

