---
title: "Multisequence alighment"
author: "Rafael Villasmil"
date: "10/16/2021"
output: pdf_document
---

```{r install-packages, echo=FALSE}
install.packages("ape")
install.packages("seqinr")
```

```{r load-packages}
library(ape)
#library(seqinr)
library(Biostrings)
library(msa)
```

```{r read-fasta-files-from-ncbi}
```

```{r load_seq_of_paper_species}
setwd("~/GitHub/cd4-aligment-nj/speciens _in_paper")
cd4_stringset_paper <- readDNAStringSet(list.files(pattern = "fa"))
cd4_stringset_paper <- readDNAStringSet(list.files(pattern = "fa"))
cd4_stringset_paper@ranges@NAMES
cd4_stringset_paper@ranges@width
cd4_ClustalW_paper <- msa(cd4_stringset_paper, method="ClustalW") 
working_clustalW <- cd4_ClustalW_paper
rownames(working_clustalW) <- c( "Mouse","Norway rat", "Human","Chimpanzee", "Rhesus monkey","Pig","Cattle", "Dog","Rabbit", "Chicken")
autoMasked <- maskGaps(working_clustalW, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")
cd4_nj <- nj(sdist)
plot(cd4_nj)
```


```{r explore cd4_ClustalW}
rownames(cd4_ClustalW)
rownames(working_clustalW) <- c("Human", "Chimpanzee", "Rhesus monkey", 
                                "Green Monkey", "Cat", "Dog", "Mouse",
                                "Norway rat", "Pig", "Cattle")
detail(cd4_ClustalW)
maskTest <- cd4_ClustalW
rowmask(maskTest) <- IRanges(start=1,end=3)
rowmask(maskTest)
maskTest
```

use Multiple Aligment Objects 
```{r distance_matrix_calculation}
autoMasked <- maskGaps(working_clustalW, min.fraction=0.5, min.block.width=4)
sdist <- stringDist(as(autoMasked,"DNAStringSet"), method="hamming")


clust <- hclust(sdist, method = "single")
plot(clust)

cd4_nj <- nj(sdist)
plot(cd4_nj)
```


https://stackoverflow.com/questions/35016713/how-do-i-import-a-jsonl-file-in-r-and-how-do-i-transform-it-in-csv
```{r}
library(jsonlite)
library(dplyr)

list_directories <- 
  list.dirs(path = "C:/Users/villasmilr/Documents/GitHub/cd4-aligment-nj/data")
jsnol_directories <- list_directories[seq(3, 41, by = 2)]

table1 <- matrix(ncol = 8)
colnames(table1) <- c("Species", "Chromosome", "Gene_length", "No.of_exon", 
                      "Exon(bp)","Coding_region_(bp)", "Intron(bp)", "UTR(bp)")
table1 <- as.data.frame(table1)
 for (i in 1:length(jsnol_directories)){
  lines <- readLines(paste(jsnol_directories[i], "/data_report.jsonl", sep= ""))
  lines <- lapply(lines, fromJSON)
  lines <- lapply(lines, unlist)
  data_report <- bind_rows(lines)
  
  gene <- readDNAStringSet(paste(jsnol_directories[i], "/gene.fna", sep= ""))
  protein <- 
    readAAStringSet(paste(jsnol_directories[i], "/protein.faa", sep= ""))
  rna <- readRNAStringSet(paste(jsnol_directories[i], "/rna.fna", sep= ""))
  
  table1[i,"Species"] <- data_report$taxname
  table1[i,"Chromosome"] <- data_report$chromosomes
  table1[i,"Gene_length"] <- gene@ranges@width[1]
  table1[i,"Coding_region_(bp)"] <- rna@ranges@width[1]
  
}


```






```{r protr_package}
library(protr)

setwd("~/GitHub/cd4-aligment-nj/data")
cd4_seq <- c(list.files(pattern = "fa"))
cd4_dna <- read.dna(list.files(pattern = "fa"))
```

